### PR TITLE
Add test for synchronous replication

### PR DESCRIPTION
**Pull request by rg2011**
This PR adds a new test for resource replication, based on [client-go FakeDynamicClient](https://pkg.go.dev/k8s.io/client-go/dynamic/fake) to simulate a connection to an actual cluster. The test checks that kube-mgmt reads the resources in this fake cluster and pushes them to a mocked OPA data interface.

Unfortunately, `FakeDynamicClient` brings in a few new dependencies (*github.com/pkg/errors*, *github.com/evanphx/json-patch*) and vendors a few more libs. But I think it's worth adding the test, to address confidently some pending issues related to resource sync that I intend to fix (#11, #134)

To ease testing, I have modified `data.New` to accept the fake `dynamic.Interface` as a parameter, instead of a `*rest.Config`, which incidentally also fixes issue #99.

I have also refactored `generateSyncPayload` and `objPath` as functions instead of methods, and made `sync.Run` receive a `context.Context` instead of returning one `quit chan struct{}` per resource, for the sake of easier testing.

### CODE REVIEW

File: pkg/data/generic.go [hunk 1]
```diff
@@ -39,35 +38,27 @@ const (
 )
 
 // New returns a new GenericSync that cna be started.
-func New(kubeconfig *rest.Config, opa opa_client.Data, ns types.ResourceType) *GenericSync {
+func New(client dynamic.Interface, opa opa_client.Data, ns types.ResourceType) *GenericSync {
```
Comment (srenatus) [inline pkg/data/generic.go:41]: I suppose we don't need to worry about compatibility here...? I.e. do we think it's common that people use this as a golang module? For them, I think, this change in function signature would be breaking stuff on upgrade. WDYT?

File: pkg/data/mock.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+func (f *mockData) PutData(path string, value interface{}) error {
+	if f.onUpdate != nil {
+		defer f.onUpdate()
+	}
+	f.data = value
+	return nil
+}
+
+// PostData implements Data. Currently not supported.
+func (f *mockData) PostData(path string, value interface{}) (json.RawMessage, error) {
```
Comment (srenatus) [inline pkg/data/mock.go:37]: [nit] 
```suggestion
func (*mockData) PostData(string, interface{}) (json.RawMessage, error) {
```

File: pkg/data/mock.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+// Prefix implements Data
+func (f *mockData) Prefix(path string) opa_client.Data {
+	f.prefix = append(f.prefix, path)
+	return f
+}
+
+// PatchData implements Data. Currently not supported.
+func (f *mockData) PatchData(path string, op string, value *interface{}) error {
```
Comment (srenatus) [inline pkg/data/mock.go:23]: [nit] 
```suggestion
func (*mockData) PatchData(string, string, *interface{}) error {
```

**Review by srenatus — COMMENTED** [review]
Some nitpicks, but nothing that can't be ignored; just stylistic preferences.

I'm uncertain about the API compatibility; do you, or does anyone, have any insights into that...?

File: pkg/data/generic.go (hunk 1 above)
Comment (rg2011) [reply pkg/data/generic.go:41]: You are right. I can add a new function `NewFromInterface`, for instance, that accepts the dynamic.Interface. And turn the function `New` into an stub that creates the interface from Config, and call `NewFromInterface`.

Comment (srenatus) [reply pkg/data/generic.go:41]: Sounds reasonable, let's do that. 👍 Thanks!

Comment (rg2011) [reply pkg/data/generic.go]: changes applied and branch rebased

**Review by srenatus — APPROVED** [review]
Thanks a lot!

Something seems to be slightly odd with your committer email vs github account settings: github still thinks your a _first-time contributor_. Is that something you care about? Otherwise, this looks good to go.

**Comment by rg2011** [comment]
Thanks for pointing out. I've reviewed my github settings and pushed again with the proper user and email (I hope). Let's see if it makes a difference. Otherwise never mind, let's go with with this PR as it is and I'll try to figure out later.

**Review by srenatus — APPROVED** [review]
Thanks

**Pull request merged**